File: go/oidc_service.go
package main

import (
	"bufio"
	"context"
	"crypto/rand"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"net"
	"net/http"
	"net/url"
	"os"
	"os/exec"
	"runtime"
	"strconv"
	"strings"
	"time"
)

type discoveryDoc struct {
	AuthorizationEndpoint string `json:"authorization_endpoint"`
}

type tokenInfo struct {
	AccessToken string
	ExpiresIn   string
	ExpiresAt   *time.Time
}

func acquireToken(authority, clientID, scope, redirectURI string, listenPort int) error {
	for {
		fmt.Printf("Getting token from: %s\n", authority)

		// Get authorization endpoint
		authEndpoint, err := getAuthEndpoint(authority)
		if err != nil {
			return err
		}

		// Generate security parameters
		state := randomString(16)
		nonce := randomString(16)

		// Build auth URL
		authURL := buildAuthURL(authEndpoint, clientID, scope, redirectURI, state, nonce)

		// Start callback server
		callbackChan := make(chan map[string]string, 1)
		server := startServer(redirectURI, listenPort, callbackChan)

		// Open browser
		fmt.Println("Opening browser...")
		if err := openBrowser(authURL); err != nil {
			fmt.Printf("Open manually: %s\n", authURL)
		}

		// Wait for callback
		var tokenResult *tokenInfo
		select {
		case params := <-callbackChan:
			tokenResult, err = processCallback(params, state)
			if err != nil {
				server.Shutdown(context.Background())
				return err
			}
		case <-time.After(5 * time.Minute):
			server.Shutdown(context.Background())
			return fmt.Errorf("timeout waiting for callback")
		}

		server.Shutdown(context.Background())

		// Display token information
		displayTokenInfo(tokenResult)

		// Interactive prompt
		if !promptForAction() {
			break // Exit the loop
		}
		// Continue the loop to refresh the token
	}

	return nil
}

func getAuthEndpoint(authority string) (string, error) {
	url := strings.TrimSuffix(authority, "/") + "/.well-known/openid-configuration"
	resp, err := http.Get(url)
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()

	var doc discoveryDoc
	if err := json.NewDecoder(resp.Body).Decode(&doc); err != nil {
		return "", err
	}

	if doc.AuthorizationEndpoint == "" {
		return "", fmt.Errorf("no authorization endpoint found")
	}

	return doc.AuthorizationEndpoint, nil
}

func buildAuthURL(endpoint, clientID, scope, redirectURI, state, nonce string) string {
	params := url.Values{
		"response_type": {"token"},
		"response_mode": {"form_post"},
		"client_id":     {clientID},
		"redirect_uri":  {redirectURI},
		"scope":         {scope},
		"state":         {state},
		"nonce":         {nonce},
		"prompt":        {"login"},
	}
	return endpoint + "?" + params.Encode()
}

func processCallback(params map[string]string, expectedState string) (*tokenInfo, error) {
	if err := params["error"]; err != "" {
		desc := params["error_description"]
		if desc != "" {
			return nil, fmt.Errorf("auth error: %s - %s", err, desc)
		}
		return nil, fmt.Errorf("auth error: %s", err)
	}

	if params["state"] != expectedState {
		return nil, fmt.Errorf("state mismatch")
	}

	token := params["access_token"]
	if token == "" {
		return nil, fmt.Errorf("no access token received")
	}

	result := &tokenInfo{
		AccessToken: token,
		ExpiresIn:   params["expires_in"],
	}

	// Calculate expiration time if available
	if expiresIn := params["expires_in"]; expiresIn != "" {
		if seconds, err := strconv.Atoi(expiresIn); err == nil {
			expires := time.Now().Add(time.Duration(seconds) * time.Second)
			result.ExpiresAt = &expires
		}
	}

	return result, nil
}

func startServer(redirectURI string, listenPort int, callbackChan chan<- map[string]string) *http.Server {
	// Extract port from redirect URI
	addr := ":5000" // default
	if u, err := url.Parse(redirectURI); err == nil && u.Host != "" {
		addr = u.Host
		if !strings.Contains(addr, ":") {
			addr += ":5000"
		}
	}

	// Override port with --listen-port if provided
	if listenPort > 0 {
		host := "localhost"
		if u, err := url.Parse(redirectURI); err == nil && u.Host != "" {
			if h, _, err := net.SplitHostPort(u.Host); err == nil {
				host = h
			} else if u.Hostname() != "" {
				host = u.Hostname()
			}
		}
		addr = fmt.Sprintf("%s:%d", host, listenPort)
	}

	server := &http.Server{
		Addr: addr,
		Handler: http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			handleCallback(w, r, callbackChan)
		}),
	}

	go server.ListenAndServe()
	return server
}

func handleCallback(w http.ResponseWriter, r *http.Request, callbackChan chan<- map[string]string) {
	params := make(map[string]string)

	// Parse POST form or GET query params
	if r.Method == "POST" {
		r.ParseForm()
		for k, v := range r.PostForm {
			if len(v) > 0 {
				params[k] = v[0]
			}
		}
	} else {
		for k, v := range r.URL.Query() {
			if len(v) > 0 {
				params[k] = v[0]
			}
		}
	}

	// Send to channel
	select {
	case callbackChan <- params:
	default:
	}

	// Response to browser
	token := params["access_token"]
	errMsg := params["error"]

	if errMsg != "" {
		fmt.Fprintf(w, "<h1>Error</h1><p>%s</p>", errMsg)
	} else if token == "" {
		fmt.Fprint(w, "<h1>Error</h1><p>No token received</p>")
	} else {
		fmt.Fprint(w, "<h1>Success!</h1><p>Token received. You can close this window.</p>")
	}
}

func randomString(length int) string {
	bytes := make([]byte, length/2)
	rand.Read(bytes)
	return hex.EncodeToString(bytes)
}

func openBrowser(url string) error {
	var cmd string
	var args []string

	switch runtime.GOOS {
	case "windows":
		cmd, args = "rundll32", []string{"url.dll,FileProtocolHandler", url}
	case "darwin":
		cmd, args = "open", []string{url}
	default:
		cmd, args = "xdg-open", []string{url}
	}

	return exec.Command(cmd, args...).Start()
}

func displayTokenInfo(token *tokenInfo) {
	fmt.Printf("\n✅ Success! Access Token: %s\n", token.AccessToken)

	if token.ExpiresAt != nil {
		fmt.Printf("Expires: %s\n", token.ExpiresAt.Format("2006-01-02 15:04:05"))
	}
}

func promptForAction() bool {
	fmt.Print("\nPress 'R' to refresh token, or 'Enter' to exit: ")

	reader := bufio.NewReader(os.Stdin)
	input, err := reader.ReadString('\n')
	if err != nil {
		return false
	}

	input = strings.TrimSpace(strings.ToUpper(input))
	return input == "R"
}

File: dotnet/OidcService.cs
using System.Diagnostics;
using System.Net;
using System.Net.Http;
using System.Text;
using System.Text.Json;
using System.Web;

namespace OidcTool;

public class OidcService
{
    private readonly HttpClient _httpClient;
    private readonly TokenCache _tokenCache;

    public OidcService(HttpClient httpClient)
    {
        _httpClient = httpClient;
        _tokenCache = new TokenCache();
    }

    public async Task AcquireTokenAsync(string authority, string clientId, string scope, string redirectUri)
    {
        try
        {
            Console.WriteLine($"Starting OIDC token acquisition...");
            Console.WriteLine($"Authority: {authority}");
            Console.WriteLine($"Client ID: {clientId}");
            Console.WriteLine($"Scope: {scope}");
            Console.WriteLine($"Redirect URI: {redirectUri}");
            Console.WriteLine();

            // Check cache first
            var cachedToken = _tokenCache.GetToken(authority, clientId, scope);
            if (cachedToken != null)
            {
                Console.WriteLine("✅ Found valid cached token!");
                if (!string.IsNullOrEmpty(cachedToken.Oid))
                {
                    Console.WriteLine($"OID: {cachedToken.Oid}");
                }
                Console.WriteLine($"Access Token: {cachedToken.AccessToken}");
                Console.WriteLine($"Expires at: {cachedToken.ExpiresAt:yyyy-MM-dd HH:mm:ss} UTC");
                Console.WriteLine($"Token Type: {cachedToken.TokenType ?? "Bearer"}");
                Console.WriteLine();
                return;
            }

            Console.WriteLine("No valid cached token found. Starting interactive authentication...");
            Console.WriteLine();

            // Discover OIDC configuration
            var discoveryDocument = await GetDiscoveryDocumentAsync(authority);
            
            if (discoveryDocument?.AuthorizationEndpoint == null)
            {
                Console.WriteLine("Error: Could not retrieve authorization endpoint from discovery document");
                return;
            }

            Console.WriteLine($"Authorization Endpoint: {discoveryDocument.AuthorizationEndpoint}");
            Console.WriteLine();

            // Generate state parameter for security
            var state = Guid.NewGuid().ToString("N");
            
            // Build authorization URL
            var authUrl = BuildAuthorizationUrl(discoveryDocument.AuthorizationEndpoint, clientId, scope, state, redirectUri);
            
            Console.WriteLine("Opening browser for authentication...");
            Console.WriteLine($"Authorization URL: {authUrl}");
            Console.WriteLine();

            // Start local HTTP listener for callback
            var callbackListener = new HttpListener();
            
            // Parse redirect URI to get the correct prefix
            try
            {
                var uri = new Uri(redirectUri);
                var prefix = $"{uri.Scheme}://{uri.Host}:{uri.Port}/";
                callbackListener.Prefixes.Add(prefix);
            }
            catch
            {
                // Fallback to default
                callbackListener.Prefixes.Add("http://localhost:5000/");
            }
            
            callbackListener.Start();

            // Open browser
            Process.Start(new ProcessStartInfo
            {
                FileName = authUrl,
                UseShellExecute = true
            });

            Console.WriteLine("Waiting for callback...");
            Console.WriteLine("Please complete the authentication in your browser.");
            Console.WriteLine();

            // Wait for callback
            var context = await callbackListener.GetContextAsync();
            var request = context.Request;
            var response = context.Response;

            // Parse callback parameters
            System.Collections.Specialized.NameValueCollection? formParams = null;
            string? returnedState = null;
            string? accessToken = null;
            string? error = null;
            string? errorDescription = null;
            string? tokenType = null;
            string? expiresIn = null;

            if (request.HttpMethod == "POST" && request.HasEntityBody)
            {
                // Read form data from POST body
                using var reader = new StreamReader(request.InputStream, request.ContentEncoding);
                var formData = await reader.ReadToEndAsync();
                formParams = HttpUtility.ParseQueryString(formData);
                
                returnedState = formParams["state"];
                accessToken = formParams["access_token"];
                error = formParams["error"];
                errorDescription = formParams["error_description"];
                tokenType = formParams["token_type"];
                expiresIn = formParams["expires_in"];
            }
            else
            {
                // Fallback to query parameters for GET requests
                var queryParams = HttpUtility.ParseQueryString(request.Url?.Query ?? string.Empty);
                returnedState = queryParams["state"];
                accessToken = queryParams["access_token"];
                error = queryParams["error"];
                errorDescription = queryParams["error_description"];
                tokenType = queryParams["token_type"];
                expiresIn = queryParams["expires_in"];
            }

            // Send response to browser
            string responseString;
            if (!string.IsNullOrEmpty(error))
            {
                responseString = $"<html><body><h1>Error</h1><p>{error}: {errorDescription}</p></body></html>";
                response.StatusCode = 400;
            }
            else if (string.IsNullOrEmpty(accessToken))
            {
                responseString = "<html><body><h1>Error</h1><p>No access token received</p></body></html>";
                response.StatusCode = 400;
            }
            else if (returnedState != state)
            {
                responseString = "<html><body><h1>Error</h1><p>Invalid state parameter</p></body></html>";
                response.StatusCode = 400;
            }
            else
            {
                responseString = $"<html><body><h1>Success</h1><p>Access token received. You can close this window.</p><div style='width:100%'><code>{accessToken}</code></div></body></html>";
                response.StatusCode = 200;
            }

            byte[] buffer = Encoding.UTF8.GetBytes(responseString);
            response.ContentLength64 = buffer.Length;
            response.OutputStream.Write(buffer, 0, buffer.Length);
            response.OutputStream.Close();
            callbackListener.Stop();

            // Handle result
            if (!string.IsNullOrEmpty(error))
            {
                Console.WriteLine($"Error: {error}");
                if (!string.IsNullOrEmpty(errorDescription))
                {
                    Console.WriteLine($"Description: {errorDescription}");
                }
                return;
            }

            if (returnedState != state)
            {
                Console.WriteLine("Error: State parameter mismatch");
                return;
            }

            if (string.IsNullOrEmpty(accessToken))
            {
                Console.WriteLine("Error: No access token received");
                return;
            }

            Console.WriteLine("✅ Access token acquired successfully!");
            Console.WriteLine($"Access Token: {accessToken}");
            Console.WriteLine();

            // Cache the token
            var expiresInSeconds = int.TryParse(expiresIn, out var expiry) ? expiry : 3600; // Default 1 hour
            var tokenCacheEntry = new TokenCacheEntry
            {
                AccessToken = accessToken,
                Authority = authority,
                ClientId = clientId,
                Scope = scope,
                TokenType = tokenType ?? "Bearer",
                ExpiresAt = DateTime.UtcNow.AddSeconds(expiresInSeconds)
            };

            _tokenCache.SaveToken(tokenCacheEntry);
            Console.WriteLine($"🔄 Token cached successfully. Expires at: {tokenCacheEntry.ExpiresAt:yyyy-MM-dd HH:mm:ss} UTC");
            Console.WriteLine();

            // Optionally decode and display token info
            await DisplayTokenInfoAsync(accessToken, discoveryDocument);
        }
        catch (Exception ex)
        {
            Console.WriteLine($"Error: {ex.Message}");
        }
    }

    private async Task<DiscoveryDocument?> GetDiscoveryDocumentAsync(string authority)
    {
        try
        {
            var discoveryUrl = $"{authority.TrimEnd('/')}/.well-known/openid-configuration";
            Console.WriteLine($"Fetching discovery document from: {discoveryUrl}");
            
            var response = await _httpClient.GetAsync(discoveryUrl);
            response.EnsureSuccessStatusCode();
            
            var json = await response.Content.ReadAsStringAsync();
            var options = new JsonSerializerOptions
            {
                PropertyNamingPolicy = JsonNamingPolicy.SnakeCaseLower
            };
            
            return JsonSerializer.Deserialize<DiscoveryDocument>(json, options);
        }
        catch (Exception ex)
        {
            Console.WriteLine($"Error fetching discovery document: {ex.Message}");
            return null;
        }
    }

    private string BuildAuthorizationUrl(string authorizationEndpoint, string clientId, string scope, string state, string redirectUri)
    {
        var queryParams = new Dictionary<string, string>
        {
            ["response_type"] = "token",
            ["response_mode"] = "form_post",
            ["client_id"] = clientId,
            ["redirect_uri"] = redirectUri,
            ["scope"] = scope,
            ["state"] = state,
            ["nonce"] = Guid.NewGuid().ToString("N")
        };

        var query = string.Join("&", queryParams.Select(kvp => 
            $"{HttpUtility.UrlEncode(kvp.Key)}={HttpUtility.UrlEncode(kvp.Value)}"));

        return $"{authorizationEndpoint}?{query}";
    }

    private async Task DisplayTokenInfoAsync(string accessToken, DiscoveryDocument discoveryDocument)
    {
        try
        {
            Console.WriteLine("Token Information:");
            Console.WriteLine($"Length: {accessToken.Length} characters");
            
            // If there's a userinfo endpoint, try to get user information
            if (!string.IsNullOrEmpty(discoveryDocument.UserinfoEndpoint))
            {
                Console.WriteLine("Fetching user information...");
                
                var request = new HttpRequestMessage(HttpMethod.Get, discoveryDocument.UserinfoEndpoint);
                request.Headers.Authorization = new System.Net.Http.Headers.AuthenticationHeaderValue("Bearer", accessToken);
                
                var response = await _httpClient.SendAsync(request);
                if (response.IsSuccessStatusCode)
                {
                    var userInfoJson = await response.Content.ReadAsStringAsync();
                    var userInfo = JsonSerializer.Deserialize<JsonElement>(userInfoJson);
                    
                    Console.WriteLine("User Information:");
                    Console.WriteLine(JsonSerializer.Serialize(userInfo, new JsonSerializerOptions { WriteIndented = true }));
                }
                else
                {
                    Console.WriteLine($"Could not fetch user info: {response.StatusCode}");
                }
            }
        }
        catch (Exception ex)
        {
            Console.WriteLine($"Error displaying token info: {ex.Message}");
        }
    }

    public void DisplayCacheInfo()
    {
        _tokenCache.DisplayCacheInfo();
    }

    public void ClearCache()
    {
        _tokenCache.ClearCache();
        Console.WriteLine("Token cache cleared successfully.");
    }

    public void RemoveTokenFromCache(string authority, string clientId, string scope)
    {
        _tokenCache.RemoveToken(authority, clientId, scope);
        Console.WriteLine("Token removed from cache.");
    }
}

public class DiscoveryDocument
{
    public string? AuthorizationEndpoint { get; set; }
    public string? TokenEndpoint { get; set; }
    public string? UserinfoEndpoint { get; set; }
    public string? JwksUri { get; set; }
    public string? Issuer { get; set; }
}
